server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -698,50 +699,123 @@ public String toString() {
                 // so this is a reasonable approximation, and the stats reported to the disk-based allocator don't hit this code path
                 // anyway since they always use IndexShard#getRecoveryState which is never transported over the wire.
-                fileDetailsComplete = fileDetails.isEmpty() == false;
+                complete = fileDetails.isEmpty() == false;
             }
-            sourceThrottlingInNanos = in.readLong();
-            targetThrottleTimeInNanos = in.readLong();
         }
 
         @Override
-        public synchronized void writeTo(StreamOutput out) throws IOException {
-            super.writeTo(out);
-            final File[] files = fileDetails.values().toArray(new File[0]);
+        public void writeTo(StreamOutput out) throws IOException {
+            final File[] files = values().toArray(new File[0]);
             out.writeVInt(files.length);
             for (File file : files) {
                 file.writeTo(out);
             }
             if (out.getVersion().onOrAfter(StoreStats.RESERVED_BYTES_VERSION)) {
-                out.writeBoolean(fileDetailsComplete);
+                out.writeBoolean(complete);
             }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            if (params.paramAsBoolean("detailed", false)) {
+                builder.startArray(Fields.DETAILS);
+                for (File file : values()) {
+                    file.toXContent(builder, params);
+                }
+                builder.endArray();
+            }
+
+            return builder;
+        }
+
+        public void addFileDetails(String name, long length, boolean reused) {
+            addFileDetails(name, new File(name, length, reused));
+        }
+
+        public void addFileDetails(String name, File file) {
+            assert complete == false : "addFileDetail for [" + name + "] when file details are already complete";
+            File existing = fileDetails.put(name, file);
+            assert existing == null : "file [" + name + "] is already reported";
+        }
+
+        public void addRecoveredBytesToFile(String name, long bytes) {
+            File file = fileDetails.get(name);
+            file.addRecoveredBytes(bytes);
+        }
+
+        public File get(String name) {
+            return fileDetails.get(name);
+        }
+
+        public void setComplete() {
+            complete = true;
+        }
+
+        public int size() {
+            return fileDetails.size();
+        }
+
+        public boolean isEmpty() {
+            return fileDetails.isEmpty();
+        }
+
+        public void clear() {
+            fileDetails.clear();
+            complete = false;
+        }
+
+        public Collection<File> values() {
+            return fileDetails.values();
+        }
+
+        public boolean complete() {
```

Comment:
We're not consistent in the codebase about this, but I'd prefer `isComplete()` as we have `setComplete()`